src/seer_server.erl

```diff
@@ -0,0 +1,166 @@
+         carbon_buffer = [] :: [carbon_batch()],
+         buffer_size = 0 :: non_neg_integer(),
+         max_buffer_size :: non_neg_integer()}).
+
+-type state() :: #state{}.
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init(_) ->
+    {ok, Host} = inet:gethostname(),
+    BinHost = case unicode:characters_to_binary(Host, latin1) of
+                Name when is_binary(Name) ->
+                    Name
+              end,
+    InitialState = #state{mode = ?ENV(?ENV_MODE, ?DEFAULT_MODE),
+                          prefix = ?ENV(?ENV_PREFIX, ?DEFAULT_PREFIX),
+                          host = BinHost,
+                          poll_interval = ?ENV(?ENV_POLL_INTERVAL, ?DEFAULT_POLL_INTERVAL),
+                          max_buffer_size = ?ENV(?ENV_MAX_BUFFER_SIZE, ?DEFAULT_MAX_BUFFER_SIZE)},
+    self() ! setup,
+    {ok, InitialState, 0}.
+
+handle_call(_, _From, State) ->
+    {reply, {error, undefined_call}, State}.
+
+handle_cast(stop, State) ->
+    {stop, normal, State};
+handle_cast(_, State) ->
+    {noreply, State}.
+
+handle_info(setup, #state{mode = Mode, poll_interval = Interval} = State) ->
+    {Msg, SetupState} = case Mode of
+                          stdout ->
+                              {poll_stdout, State};
+                          carbon ->
+                              carbon_setup(State)
+                        end,
+    timer:send_after(Interval, Msg),
+    {noreply, SetupState};
+handle_info(poll_stdout, #state{poll_interval = Interval} = State) ->
+    Metrics = seer:read_all(),
+    io:format("~w~n", [Metrics]),
+    timer:send_after(Interval, poll_stdout),
+    {noreply, State};
+handle_info(poll_carbon,
+            #state{prefix = Prefix, host = Host, tcp_socket = Socket, poll_interval = Interval} = State) ->
+    Metrics = seer:read_all(),
+    CarbonStrings = seer_utils:carbon_format(Prefix, Host, Metrics),
+    case carbon_send_batch(Socket, CarbonStrings) of
+      ok ->
+          timer:send_after(Interval, poll_carbon),
+          {noreply, State};
+      {error, UnsentStrings} ->
+          timer:send_after(Interval, poll_carbon_offline),
+          timer:send_after(?TCP_RECONNECTION_INTERVAL, carbon_reconnect),
+          {noreply, State#state{mode = carbon_offline, carbon_buffer = [UnsentStrings]}}
+    end;
+handle_info(poll_carbon_offline, #state{mode = carbon_reconnected} = State) ->
+    self() ! poll_carbon,
+    {noreply, State#state{mode = carbon}};
+handle_info(poll_carbon_offline,
+            #state{prefix = Prefix,
+                   host = Host,
+                   poll_interval = Interval,
+                   carbon_buffer = Buffer,
+                   buffer_size = BufferSize,
+                   max_buffer_size = MaxBufferSize} =
+                State) ->
+    Metrics = seer:read_all(),
+    CarbonStrings = seer_utils:carbon_format(Prefix, Host, Metrics),
+    {NewBuffer, NewBufferSize} = case BufferSize < MaxBufferSize of
```

Comment:
Maybe the following is clearer? Though I will admit it doesn't handle values greater than `MaxBufferSize` very well.

```erlang
case BufferSize of
    MaxBufferSize -> ...;
    _ -> ...
end
```

Reply:
While in principle, since the `MaxBufferSize` case is handled, it should never go higher than that, but I have a strong aversion to check only the boundaries and not further. Would something like
```
case BufferSize of
    BuffSize when BuffSize < MaxBufferSize ->
        {[CarbonStrings | Buffer], BuffSize + 1};
    BuffSize ->
        {[CarbonStrings | lists:droplast(Buffer)], BuffSize}
end
```
be better?

Reply:
Nope! I prefer the code you currently have :)